core/src/main/java/com/yandex/yoctodb/util/immutable/impl/TrieByteArraySortedSet.java

```diff
@@ -0,0 +1,445 @@
+        while (true) {
+            final int metadata = Byte.toUnsignedInt(nodes.get(movingOffset));
+            movingOffset += Byte.BYTES;
+
+            if (hasInfix(metadata)) { // there is an infix
+                int infixSize = nodes.getInt(movingOffset);
+                movingOffset += Integer.BYTES;
+                movingOffset += infixSize * Byte.BYTES;
+            }
+
+            if (hasValue(metadata)) { // there is a value
+                return nodes.getInt(movingOffset);
+            }
+
+            switch (edgeType(metadata)) {
+                case EDGES_BITSET:
+                    int min = Byte.toUnsignedInt(nodes.get(movingOffset++));
+                    int max = Byte.toUnsignedInt(nodes.get(movingOffset++));
+
+                    movingOffset += BufferBitSet.arraySize(max - min + 1) * Long.BYTES;
+                    movingOffset = nodes.getLong(movingOffset);
+                    break;
+                case EDGES_CONDENSED:
+                    movingOffset += 2 * Byte.BYTES;
+                    movingOffset = nodes.getLong(movingOffset);
+                    break;
+                default: // should be unreachable
+            }
+        }
+    }
+
+    private int takeLastValueOnRight(long movingOffset) {
+        while (true) {
+            final int metadata = Byte.toUnsignedInt(nodes.get(movingOffset));
+            movingOffset += Byte.BYTES;
+
+            if (hasInfix(metadata)) { // there is an infix
+                int infixSize = nodes.getInt(movingOffset);
+                movingOffset += Integer.BYTES;
+                movingOffset += infixSize * Byte.BYTES;
+            }
+
+            int value = NOT_FOUND;
+            if (hasValue(metadata)) { // there is a value
+                value = nodes.getInt(movingOffset);
+                movingOffset += Integer.BYTES;
+            }
+
+            switch (edgeType(metadata)) {
+                case EDGES_SINGLE:
+                    movingOffset += Byte.BYTES;
+                    movingOffset = nodes.getLong(movingOffset);
+                    break;
+                case EDGES_BITSET: {
+                    int min = Byte.toUnsignedInt(nodes.get(movingOffset++));
+                    int max = Byte.toUnsignedInt(nodes.get(movingOffset++));
+
+                    int index = BufferBitSet.cardinalityTo(nodes, movingOffset, max - min);
+                    movingOffset += BufferBitSet.arraySize(max - min + 1) * Long.BYTES;
+                    movingOffset = nodes.getLong(movingOffset + index * Long.BYTES);
+                    break;
+                }
+                case EDGES_CONDENSED: {
+                    int min = Byte.toUnsignedInt(nodes.get(movingOffset++));
+                    int max = Byte.toUnsignedInt(nodes.get(movingOffset++));
+
+                    movingOffset = nodes.getLong(movingOffset + (max - min) * Long.BYTES);
+                    break;
+                }
+                case EDGES_NONE:
```

Comment:
default is undefined

Reply:
It's redundant because `edgeType(int metadata)` is a pure function that may emit only 4 distinct values which have covered by all the cases in this switch.

If you insist, I can change `case EDGES_NONE:` for `default:`, but I don't think that would be any way better for readability.

Reply:
Anyway, forget about this. Added empty default clause

Reply:
Empty `default` case is quite a risk when maintaining the code.

Reply:
That's interesting. I thought that an empty default case is meaningless for a program flow and Java would treat it as if it's not here, and it's only indicating for the code reader that I have thought about that option and leaved it with exactly that implementation. Could you please extend your opinion on that?

Reply:
Just imaging extending the `enum` when somebody adds a feature in future. `default` empty branch will silently swallow the case which may lead to unexpected results.